src/sentry/api/endpoints/organization_dashboard_details.py

```diff
@@ -1,11 +1,97 @@
 from __future__ import absolute_import
 
+from django.db.models import Max
+from rest_framework import serializers
+from rest_framework.response import Response
+
 from sentry.api.base import DocSection
 from sentry.api.bases.dashboard import (
     OrganizationDashboardEndpoint
 )
 from sentry.api.serializers import serialize
-from sentry.models import ObjectStatus
+from sentry.api.serializers.rest_framework import ListField, ValidationError
+from sentry.models import ObjectStatus, Widget
+
+
+def get_next_dashboard_order(dashboard_id):
+    max_order = Widget.objects.filter(
+        dashboard_id=dashboard_id,
+    ).aggregate(Max('order'))
+    return max_order['order__max'] + 1
+
+
+def remove_widgets(dashboard_widgets, widget_data):
+    """
+    Removes current widgets belonging to dashboard not in widget_data.
+    Returns remaining widgets.
+    """
+    widget_titles = [wd['id'] for wd in widget_data]
+    dashboard_widgets.exclude(
+        id__in=widget_titles
+    ).update(status=ObjectStatus.PENDING_DELETION)
+
+    return dashboard_widgets.filter(
+        id__in=widget_titles
+    )
+
+
+def reorder_widgets(dashboard_id, widget_data):
+    """
+    Reorders Widgets given the relative order desired,
+    reorders widgets in the next possible set of numbers
+    i.e if order of widgets is 1, 2, 3
+        the reordered widgets will have order 4, 5, 6
+    """
+    dashboard_widgets = Widget.objects.filter(
+        dashboard_id=dashboard_id,
+    )
+    dashboard_widgets = list(remove_widgets(dashboard_widgets, widget_data))
+
+    # dashboard_widgets and widget_data should now have the same widgets
+    widget_data.sort(key=lambda x: x['order'])
+
+    next_order = get_next_dashboard_order(dashboard_id)
```

Comment:
Do the PUT requests get the full set of widgets in the request? If so you can use the order of elements in the PUT data to generate the ordering unless the user is sending an explicit order field as well.

Reply:
@markstory Right now I'm not requiring that they send all widgets. If a dashboard has widgets 1-5 and they send me only 2 and 3, then I will delete 1,4, and 5. So the widgets given aren't necessarily the full set to allow for deleting widgets.